From 3703a0478b09471a2ea2f68786f88beb316eebd7 Mon Sep 17 00:00:00 2001
From: rohit <rohit.sharma@r2p.com>
Date: Wed, 8 Nov 2023 11:14:41 +0000
Subject: [PATCH] fix device tree for olimex a20 som carrier board

Configures the A20 SOM Carrier Board with:
- Additional UART ports (uart2, uart3, uart4)
- Audio output via TPA3112D1PWP amplifier using codec's PA GPIO
- Microphone input via internal codec
- CSI0 camera interface
- Removes HDMI (not present on carrier board)

Audio amplifier notes (TPA3112D1PWP):
- PE5 controls shutdown via BSS138 MOSFET:
  PE5 HIGH = MOSFET on, SD pulled low = shutdown
  PE5 LOW = R30 pulls SD to VIN = enabled
- PE7 controls GAIN0, PE8 controls GAIN1
  Both set LOW for 20dB gain

Uses sun4i-codec's built-in PA GPIO support (allwinner,pa-gpios)
for proper DAPM integration - amp enables when Speaker widget is active.
---
 arch/arm/boot/dts/sun7i-a20-olinuxino-micro.dts | 85 ++++++++++++++++--------
 1 file changed, 63 insertions(+), 22 deletions(-)

diff --git a/arch/arm/boot/dts/sun7i-a20-olinuxino-micro.dts b/arch/arm/boot/dts/sun7i-a20-olinuxino-micro.dts
index 359bd0d5b3b1..a1b2c3d4e5f6 100644
--- a/arch/arm/boot/dts/sun7i-a20-olinuxino-micro.dts
+++ b/arch/arm/boot/dts/sun7i-a20-olinuxino-micro.dts
@@ -57,7 +57,10 @@
 	aliases {
 		serial0 = &uart0;
 		serial1 = &uart6;
-		serial2 = &uart7;
+		serial2 = &uart2;
+		serial3 = &uart4;
+		serial4 = &uart3;
+		serial5 = &uart7;
 		spi0 = &spi1;
 		spi1 = &spi2;
 	};
@@ -66,17 +69,6 @@
 		stdout-path = "serial0:115200n8";
 	};
 
-	hdmi-connector {
-		compatible = "hdmi-connector";
-		type = "a";
-
-		port {
-			hdmi_con_in: endpoint {
-				remote-endpoint = <&hdmi_out_con>;
-			};
-		};
-	};
-
 	leds {
 		compatible = "gpio-leds";
 		pinctrl-names = "default";
@@ -88,6 +80,29 @@
 			default-state = "on";
 		};
 	};
+
+	/*
+	 * Audio amplifier gain control (TPA3112D1PWP)
+	 * Using gpio-leds to hold GAIN pins LOW for 20dB gain
+	 * GAIN1=0, GAIN0=0 -> 20dB
+	 */
+	audio-gain-pins {
+		compatible = "gpio-leds";
+		pinctrl-names = "default";
+		pinctrl-0 = <&audio_gain_pins>;
+
+		gain0 {
+			label = "audio-gain0";
+			gpios = <&pio 4 7 GPIO_ACTIVE_HIGH>;
+			default-state = "off";
+		};
+
+		gain1 {
+			label = "audio-gain1";
+			gpios = <&pio 4 8 GPIO_ACTIVE_HIGH>;
+			default-state = "off";
+		};
+	};
 };
 
 &ahci {
@@ -96,6 +111,13 @@
 };
 
 &codec {
+	/* TPA3112D1PWP: PE5 via BSS138 MOSFET - active-low enables amp */
+	allwinner,pa-gpios = <&pio 4 5 GPIO_ACTIVE_LOW>;
+	allwinner,audio-routing =
+		"Speaker", "HP Left",
+		"Speaker", "HP Right",
+		"MIC1", "Mic Bias",
+		"Mic Bias", "Microphone";
 	status = "okay";
 };
 
@@ -123,16 +145,6 @@
 	status = "okay";
 };
 
-&hdmi {
-	status = "okay";
-};
-
-&hdmi_out {
-	hdmi_out_con: endpoint {
-		remote-endpoint = <&hdmi_con_in>;
-	};
-};
-
 &i2c0 {
 	status = "okay";
 
@@ -254,6 +266,12 @@
 		function = "gpio_out";
 		drive-strength = <20>;
 	};
+
+	audio_gain_pins: audio-gain-pins {
+		pins = "PE7", "PE8";
+		function = "gpio_out";
+		drive-strength = <20>;
+	};
 };
 
 #include "axp209.dtsi"
@@ -323,6 +341,26 @@
 	status = "okay";
 };
 
+&uart2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart2_pi_pins>, <&uart2_cts_rts_pi_pins>;
+	uart-has-rtscts;
+	status = "okay";
+};
+
+&uart3 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart3_pg_pins>, <&uart3_cts_rts_pg_pins>;
+	uart-has-rtscts;
+	status = "okay";
+};
+
+&uart4 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&uart4_pg_pins>;
+	status = "okay";
+};
+
 &uart6 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&uart6_pi_pins>;
@@ -335,6 +373,12 @@
 	status = "okay";
 };
 
+&csi0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&csi0_8bits_pins>;
+	status = "okay";
+};
+
 &usb_otg {
 	dr_mode = "otg";
 	status = "okay";
-- 
2.42.1

